Fix hidden glob reexports 146164

sitetester opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Fixes: #[146164](https://github.com/rust-lang/rust/issues/146164)

The [hidden_glob_reexports](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/builtin/static.HIDDEN_GLOB_REEXPORTS.html) lint will now skip generating a warning when the imported type points to the same type as public glob reexport

Let's create the following files to check the current behaviour 

### Current behaviour  
- fix/before/allow_when_same.rs
```rust
#![no_std] // Allows direct access to core OR Absolute path to core
#![allow(hidden_glob_reexports)]

pub use core::option::*;
use ::core::option::Option;

fn main() {
    // warning suppressed because of `allow`
    let _x: Option<i32> = Some(42);
}
```
- fix/before/warn_when_same.rs
```rust
#![warn(hidden_glob_reexports)]

pub use std::option::*;
use ::std::option::Option;

fn main() {
    // Warns when same (current buggy behavior)
    // Even though importing the same item & there is no "shadowing"


    // Should Only warn when a different/private item actually shadows the glob reexport
    let _x: Option<i32> = Some(42);
}

```
Can test using 
```bash
~/rust/build/host/stage1/bin/rustc --edition 2021 fix/before/allow_when_same.rs
```

<img width="831" height="65" alt="img_9" src="https://github.com/user-attachments/assets/f9233cf4-0220-40b2-9903-5f83f3fbd52f" />

```bash
~/rust/build/host/stage1/bin/rustc --edition 2021 fix/before/warn_when_same.rs 
```
<img width="964" height="569" alt="img_8" src="https://github.com/user-attachments/assets/cab632e1-2b8e-448c-aa58-8a6d2b158d01" />

### Fix
- Update `compiler/rustc_resolve/src/imports.rs`
- Rebuild
```bash
./x build --stage 1 # Compiler + standard library + sysroot (complete setup)
```

Create these files
- fix/after/not_warn_when_same.rs
```rust
#![warn(hidden_glob_reexports)]

pub use std::option::*;
use ::std::option::Option;

fn main() {
    // Should be no warning after fix
    let _x: Option<i32> = Some(42);
}

```
```bash
~/rust/build/host/stage1/bin/rustc --edition 2021 fix/after/not_warn_when_same.rs
```
<img width="953" height="86" alt="img_10" src="https://github.com/user-attachments/assets/15414e7e-69c2-4750-a859-f5c21f6e29db" />

- fix/after/should_warn_when_private_shadows_glob.rs
```rust
#![deny(hidden_glob_reexports)]

pub mod upstream {
    mod inner { pub struct Foo {} pub struct Bar {} }
    pub use self::inner::*;
    struct Foo {} // private item shadows `inner::Foo`
}

// mod downstream {
//     fn test() {
//         let _ = crate::upstream::Foo; // inaccessible
//     }
// }

pub fn main() {}
```
```bash
~/rust/build/host/stage1/bin/rustc --edition 2021 fix/after/should_warn_when_private_shadows_glob.rs
```
<img width="1157" height="948" alt="img_11" src="https://github.com/user-attachments/assets/c3e898d1-9374-4e11-bc96-b58827f3ba2e" />

For development/quick testing,
hidden-glob-reexports-issue-146164.rs is the official test that goes into Rust, as every bug fix needs a test as per Rust standards. It's part of Rust's test suite that runs automatically

```bash
~/rust/build/host/stage1/bin/rustc ~/rust/tests/ui/lint/hidden-glob-reexports-issue-146164.rs
```
<img width="1174" height="72" alt="image" src="https://github.com/user-attachments/assets/e9b79d7e-fe69-4a88-b452-5a2ce030789e" />

For official/CI testing
```bash
cd ~/rust # Must be in rust directory first  
./x test tests/ui/lint/hidden-glob-reexports-issue-146164.rs  # Use relative path
```

```bash
cd ~/rust 
./x test tests/ui/lint/ --test-args="146164"
```

sitetester committed:
Fix false positive in hidden_glob_reexports lint

Fixes #146164

The lint was incorrectly warning when importing the same item that's
already available through a glob re-export. This adds a check to only
warn when the explicit import and glob import resolve to different items.

sitetester committed:
Add //@ check-pass, as now code should compile successfully without any warnings or errors

sitetester committed:
Remove redundant comment

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17512883630/job/49746816939) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49746816939) [(plain)](https://github.com/rust-lang/rust/commit/e802928376da0a8f1760c9510660df3c5fe046bb/checks/49746816939/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6371 files
tidy check
tidy error: file `tests/ui/lint/issue-146164.rs` must begin with a descriptive name, consider `{reason}-issue-146164.rs`
Checking tidy rustdoc_json...
No error code explanation was removed!
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
---
Running eslint on rustdoc JS files
info: ES-Check: there were no ES version matching errors!  🎉
some tidy checks failed
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/npm --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1548:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1281:29

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:03:11
  local time: Sat Sep  6 09:41:20 UTC 2025
  network time: Sat, 06 Sep 2025 09:41:20 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

sitetester committed:
Fix false positive in hidden_glob_reexports lint

The lint was incorrectly warning when importing the same item that's
already available through a glob re-export. This adds a check to only
warn when the explicit import and glob import resolve to different items.

petrochenkov commented:
In the example
```rust
// mylib
pub use std::option::*;
use std::option::Option;
```
the lint `hidden_glob_reexports` should still warn because the private `Option` from single import will still make the public `Option` from glob import unavailable, and `mylib::Option` will result in a privacy error.

I'll close https://github.com/rust-lang/rust/issues/146164.
